File: database/app/routers/conversations.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session as OrmSession

from app.database import get_db
from app.deps import get_current_user
from app.models import Conversation, Prompt, User
from app.schemas import (
    ConversationCreateResponse,
    ConversationRead,
    PromptCreate,
    PromptCreateResponse,
    PromptRead,
    PromptResponseCreate,
)
from app.security import utcnow
from app.storage import append_transcript, save_media, transcript_path

router = APIRouter(tags=["conversations"])


def _owned_conversation(conversation_id: int, user: User, db: OrmSession) -> Conversation:
    conversation = (
        db.query(Conversation)
        .filter(Conversation.conversation_id == conversation_id, Conversation.user_id == user.id)
        .first()
    )
    if conversation is None:
        # 404 rather than 403 so we don't reveal that the id exists.
        raise HTTPException(status_code=404, detail="Conversation not found")
    return conversation


@router.post("/conversations", response_model=ConversationCreateResponse, status_code=status.HTTP_201_CREATED)
def create_conversation(
    current_user: User = Depends(get_current_user),
    db: OrmSession = Depends(get_db),
):
    conversation = Conversation(user_id=current_user.id, conversation=None, cost=0.0)
    db.add(conversation)
    db.commit()
    db.refresh(conversation)

    path = transcript_path(conversation.conversation_id)
    path.write_text("", encoding="utf-8")
    conversation.conversation = str(path)
    db.commit()
    return ConversationCreateResponse(conversation_id=conversation.conversation_id)


@router.get("/conversations", response_model=list[ConversationRead])
def list_conversations(
    current_user: User = Depends(get_current_user),
    db: OrmSession = Depends(get_db),
):
    return (
        db.query(Conversation)
        .filter(Conversation.user_id == current_user.id)
        .order_by(Conversation.creation_date)
        .all()
    )


@router.get("/conversations/{conversation_id}", response_model=ConversationRead)
def get_conversation(
    conversation_id: int,
    current_user: User = Depends(get_current_user),
    db: OrmSession = Depends(get_db),
):
    return _owned_conversation(conversation_id, current_user, db)


@router.get("/conversations/{conversation_id}/prompts", response_model=list[PromptRead])
def get_prompts(
    conversation_id: int,
    current_user: User = Depends(get_current_user),
    db: OrmSession = Depends(get_db),
):
    _owned_conversation(conversation_id, current_user, db)
    return (
        db.query(Prompt)
        .filter(Prompt.conversation_id == conversation_id)
        .order_by(Prompt.prompt_time)
        .all()
    )


@router.post(
    "/conversations/{conversation_id}/prompts",
    response_model=PromptCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
def add_prompt(
    conversation_id: int,
    payload: PromptCreate,
    current_user: User = Depends(get_current_user),
    db: OrmSession = Depends(get_db),
):
    _owned_conversation(conversation_id, current_user, db)
    if not payload.content:
        raise HTTPException(status_code=400, detail="Prompt content cannot be empty")

    input_media_path = None
    if payload.input_media_base64:
        input_media_path = save_media(
            conversation_id, payload.input_media_filename or "input", payload.input_media_base64
        )

    prompt = Prompt(
        conversation_id=conversation_id,
        content=payload.content,
        response=None,
        input_media=input_media_path,
        output_media=None,
    )
    db.add(prompt)
    db.commit()
    db.refresh(prompt)
    append_transcript(conversation_id, f"[{utcnow().isoformat()}] USER: {payload.content}")
    return PromptCreateResponse(prompt_id=prompt.prompt_id)


@router.post("/prompts/{prompt_id}/response")
def add_response(
    prompt_id: int,
    payload: PromptResponseCreate,
    current_user: User = Depends(get_current_user),
    db: OrmSession = Depends(get_db),
):
    prompt = db.query(Prompt).filter(Prompt.prompt_id == prompt_id).first()
    if prompt is None:
        raise HTTPException(status_code=404, detail="Prompt not found")
    _owned_conversation(prompt.conversation_id, current_user, db)

    if not payload.response:
        raise HTTPException(status_code=400, detail="Response content cannot be empty")

    output_media_path = None
    if payload.output_media_base64:
        output_media_path = save_media(
            prompt.conversation_id, payload.output_media_filename or "output", payload.output_media_base64
        )

    prompt.response = payload.response
    prompt.output_media = output_media_path
    db.commit()
    append_transcript(prompt.conversation_id, f"[{utcnow().isoformat()}] AI: {payload.response}")

    conversation = (
        db.query(Conversation).filter(Conversation.conversation_id == prompt.conversation_id).first()
    )
    conversation.cost = conversation.cost + payload.cost
    db.commit()
    return {"ok": True}

File: frontend/conversations.js
// /conversations, /prompts -- mirrors ../database/app/routers/conversations.py.
const express = require("express");
const fs = require("node:fs");
const { db } = require("./db");
const { ApiError } = require("./errors");
const { requireAuth } = require("./auth");
const { transcriptPath, appendTranscript, saveMedia } = require("./storage");

const router = express.Router();

// 404 rather than 403 so we don't reveal that the id exists.
function ownedConversation(conversationId, userId) {
    const conversation = db
        .prepare("SELECT * FROM conversation WHERE conversation_id = ? AND user_id = ?")
        .get(conversationId, userId);
    if (!conversation) throw new ApiError(404, "Conversation not found");
    return conversation;
}

router.post("/conversations", requireAuth, (req, res) => {
    const info = db
        .prepare("INSERT INTO conversation (user_id, conversation, creation_date, cost) VALUES (?, NULL, ?, 0)")
        .run(req.user.id, new Date().toISOString());
    const conversationId = Number(info.lastInsertRowid);

    const path = transcriptPath(conversationId);
    fs.writeFileSync(path, "", "utf8");
    db.prepare("UPDATE conversation SET conversation = ? WHERE conversation_id = ?").run(path, conversationId);

    res.status(201).json({ conversation_id: conversationId });
});

router.get("/conversations", requireAuth, (req, res) => {
    const rows = db
        .prepare("SELECT * FROM conversation WHERE user_id = ? ORDER BY creation_date")
        .all(req.user.id);
    res.json(rows);
});

router.get("/conversations/:id", requireAuth, (req, res) => {
    res.json(ownedConversation(Number(req.params.id), req.user.id));
});

router.get("/conversations/:id/prompts", requireAuth, (req, res) => {
    ownedConversation(Number(req.params.id), req.user.id);
    const rows = db
        .prepare("SELECT * FROM prompts WHERE conversation_id = ? ORDER BY prompt_time")
        .all(Number(req.params.id));
    res.json(rows);
});

router.post("/conversations/:id/prompts", requireAuth, (req, res) => {
    const conversationId = Number(req.params.id);
    ownedConversation(conversationId, req.user.id);

    const content = req.body?.content;
    if (!content) throw new ApiError(400, "Prompt content cannot be empty");

    let inputMediaPath = null;
    if (req.body?.input_media_base64) {
        inputMediaPath = saveMedia(conversationId, req.body.input_media_filename || "input", req.body.input_media_base64);
    }

    const info = db
        .prepare(
            "INSERT INTO prompts (conversation_id, prompt_time, content, response, input_media, output_media) VALUES (?, ?, ?, NULL, ?, NULL)"
        )
        .run(conversationId, new Date().toISOString(), content, inputMediaPath);

    appendTranscript(conversationId, `[${new Date().toISOString()}] USER: ${content}`);
    res.status(201).json({ prompt_id: Number(info.lastInsertRowid) });
});

router.post("/prompts/:id/response", requireAuth, (req, res) => {
    const promptId = Number(req.params.id);
    const prompt = db.prepare("SELECT * FROM prompts WHERE prompt_id = ?").get(promptId);
    if (!prompt) throw new ApiError(404, "Prompt not found");
    ownedConversation(prompt.conversation_id, req.user.id);

    const response = req.body?.response;
    if (!response) throw new ApiError(400, "Response content cannot be empty");

    let outputMediaPath = null;
    if (req.body?.output_media_base64) {
        outputMediaPath = saveMedia(
            prompt.conversation_id,
            req.body.output_media_filename || "output",
            req.body.output_media_base64
        );
    }

    db.prepare("UPDATE prompts SET response = ?, output_media = ? WHERE prompt_id = ?").run(
        response,
        outputMediaPath,
        promptId
    );
    appendTranscript(prompt.conversation_id, `[${new Date().toISOString()}] AI: ${response}`);

    const cost = Number(req.body?.cost || 0);
    db.prepare("UPDATE conversation SET cost = cost + ? WHERE conversation_id = ?").run(cost, prompt.conversation_id);

    res.json({ ok: true });
});

module.exports = { router };
